File: hs/35.hs
rotate n i = read (take (length s) $ drop i $ cycle s) :: Int
    where s = show n

fs = floor . sqrt . fromIntegral

prime n | even n    = False
        | otherwise = notElem 0 $ map (\m -> mod n m) [3,5..fs n]

solution n | not $ prime n = False
           | otherwise = all prime $
                 map (\i -> rotate n i) [1..(length $ show n) - 1]

main = print $ length $ filter solution [1..1000000]

File: py/35.py
from math import sqrt

def prime(n):
    if n % 2 == 0:
        return False

    for i in range(3, int(sqrt(n)) + 1):
        if n != i and n % i == 0:
            return False

    return True

def rotate(n, r):
    s = str(n)
    return int(s[r:] + s[:r])

def circular(n):
    return all(map(prime, [rotate(n, r) for r in range(len(str(n)))]))

print len(filter(circular, range(1000000)))
